src/middleware/axum.rs

```diff
@@ -0,0 +1,209 @@
+// Originally retired from davidB/tracing-opentelemetry-instrumentation-sdk
+// https://github.com/davidB/tracing-opentelemetry-instrumentation-sdk/blob/d3609ac2cc699d3a24fbf89754053cc8e938e3bf/axum-tracing-opentelemetry/src/middleware/trace_extractor.rs#L53
+// which is licensed under CC0 1.0 Universal
+// https://github.com/davidB/tracing-opentelemetry-instrumentation-sdk/blob/d3609ac2cc699d3a24fbf89754053cc8e938e3bf/LICENSE
+
+use axum::extract::MatchedPath;
+use http::{Request, Response};
+use pin_project_lite::pin_project;
+use std::{
+    error::Error,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{Layer, Service};
+use tracing::Span;
+use tracing_opentelemetry_instrumentation_sdk::http as otel_http;
+
+pub type Filter = fn(&str) -> bool;
+
+#[derive(Default, Debug, Clone)]
+pub struct OtelAxumLayer {
+    filter: Option<Filter>,
+}
+
+// add a builder like api
+impl OtelAxumLayer {
+    #[must_use]
+    pub fn filter(self, filter: Filter) -> Self {
+        OtelAxumLayer {
+            filter: Some(filter),
+        }
+    }
+}
+
+impl<S> Layer<S> for OtelAxumLayer {
+    /// The wrapped service
+    type Service = OtelAxumService<S>;
+    fn layer(&self, inner: S) -> Self::Service {
+        OtelAxumService {
+            inner,
+            filter: self.filter,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct OtelAxumService<S> {
+    inner: S,
+    filter: Option<Filter>,
+}
+
+impl<S, B, B2> Service<Request<B>> for OtelAxumService<S>
+where
+    S: Service<Request<B>, Response = Response<B2>> + Clone + Send + 'static,
+    S::Error: Error + 'static, //fmt::Display + 'static,
+    S::Future: Send + 'static,
+    B: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    // #[allow(clippy::type_complexity)]
+    // type Future = futures_core::future::BoxFuture<'static, Result<Self::Response, Self::Error>>;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
```

Comment:
`map_err` seems redundant here since `self.inner.poll_ready` returns the same `S::Error`:

```suggestion
        self.inner.poll_ready(cx)
```

Reply:
Hm... but now I'm reviewing davidB's code...